File: src/examples/python/mergesort.py
# Merge Sort
import math

unsorted = [7, 3, 2, 6, 8, 5, 1, 4]

node_style = {
    'shape': 'rect',
    'size': (30, 15),
    'fixed': True
}

def node_pos(i, level):
    x = (i - (len(unsorted) - 1) / 2) * 80
    y = -(level - math.ceil(math.log2(len(unsorted))) / 2) * 120
    return (x, y)

def node_id(i, level):
    return str(i) + '_' + str(level)

def animate_merge_edge(node_prev, node_cur, number, color):
    edge = (node_prev, node_cur)

    canvas.node(node_prev).color(color)
    canvas.edge(edge).add().directed(True)
    canvas.pause(0.5)

    canvas.edge(edge).duration(0.8).traverse(color).pause(0.2)
    canvas.node(node_cur).label().text(number).visible(True)
    canvas.pause(0.5)


def animate_merge(left, right, index, level):
    total_len = len(left) + len(right)
    prev_ids = [node_id(index + i, level - 1) for i in range(total_len)]
    cur_ids = [node_id(index + i, level) for i in range(total_len)]

    canvas.nodes(cur_ids).add(
        node_style,
        pos=lambda _, i: node_pos(index + i, level),
        labels={0: {'visible': False }}
    )

    canvas.edges([(cur_ids[i], cur_ids[i + 1])
        for i in range(len(cur_ids) - 1)]).add()

    canvas.pause(0.8)

    li = 0
    ri = 0
    result = []
    for i in range(total_len):
        if ri == len(right) or (li < len(left) and left[li] < right[ri]):
            result.append(left[li])
            animate_merge_edge(prev_ids[li], cur_ids[i],
                left[li], 'green')
            li += 1
        else:
            result.append(right[ri])
            animate_merge_edge(prev_ids[len(left) + ri], cur_ids[i],
                right[ri], 'red')
            ri += 1

    return result

def merge_sort(array, index=0, level=None):
    level = math.ceil(math.log2(len(array))) if level is None else level

    if len(array) == 1:
        if level != 0:
            animate_merge(array, [], index, level)
        return array

    split_index = len(array) // 2
    left = merge_sort(array[0:split_index], index, level - 1)
    right = merge_sort(array[split_index:], index + split_index, level - 1)

    return animate_merge(left, right, index, level)

canvas.pause(0.1)
top_ids = [node_id(i, 0) for i in range(len(unsorted))]

canvas.nodes(top_ids).data(unsorted).add(
    node_style,
    pos=lambda _, i: node_pos(i, 0),
    labels=lambda n: {0: {'text': n }}
)

canvas.pause(1)
merge_sort(unsorted)

File: src/examples/javascript/mergesort.js
// Merge Sort

const unsorted = [7, 3, 2, 6, 8, 5, 1, 4];

const nodeStyle = {
    shape: 'rect',
    size: [30, 15],
    fixed: true,
};

function nodePos(i, level) {
    const x = (i - (unsorted.length - 1) / 2) * 80;
    const y = -(level - Math.ceil(Math.log2(unsorted.length)) / 2) * 120;
    return [x, y];
}
function nodeId(i, level) {
    return String(i) + '_' + String(level);
}

function animateMergeEdge(nodePrev, nodeCur, number, color) {
    const edge = [nodePrev, nodeCur];

    canvas.node(nodePrev).color(color);
    canvas.edge(edge).add().directed(true);
    canvas.pause(0.5);

    canvas.edge(edge).traverse().duration(0.8).color(color).pause(0.2);
    canvas.node(nodeCur).label().text(number).visible(true);
    canvas.pause(0.5);
}

function animateMerge(left, right, index, level) {
    const totalLen = left.length + right.length;
    const prevIds = Array(totalLen)
        .fill(0)
        .map((_, i) => nodeId(index + i, level - 1));
    const curIds = Array(totalLen)
        .fill(0)
        .map((_, i) => nodeId(index + i, level));

    canvas
        .nodes(curIds)
        .add()
        .set(nodeStyle)
        .pos((n, i) => nodePos(index + i, level))
        .label()
        .visible(false);

    canvas
        .edges(
            Array(totalLen - 1)
                .fill(0)
                .map((_, i) => [curIds[i], curIds[i + 1]])
        )
        .add();

    canvas.pause(0.8);

    let li = 0;
    let ri = 0;
    let result = [];
    for (let i = 0; i < totalLen; i++) {
        if (ri === right.length || (li < left.length && left[li] < right[ri])) {
            result.push(left[li]);
            animateMergeEdge(prevIds[li], curIds[i], left[li], 'green');
            li += 1;
        } else {
            result.push(right[ri]);
            animateMergeEdge(prevIds[left.length + ri], curIds[i], right[ri], 'red');
            ri += 1;
        }
    }
    return result;
}

function mergeSort(array, index = 0, level = null) {
    level = level === null ? Math.ceil(Math.log2(array.length)) : level;

    if (array.length == 1) {
        if (level !== 0) animateMerge(array, [], index, level);
        return array;
    }
    const splitIndex = Math.floor(array.length / 2);
    const left = mergeSort(array.slice(0, splitIndex), index, level - 1);
    const right = mergeSort(array.slice(splitIndex), index + splitIndex, level - 1);

    return animateMerge(left, right, index, level);
}

canvas.pause(0.1);
const topIds = Array(unsorted.length)
    .fill(0)
    .map((_, i) => nodeId(i, 0));
canvas
    .nodes(topIds)
    .add()
    .set(nodeStyle)
    .pos((n, i) => nodePos(i, 0))
    .data(unsorted)
    .label()
    .text((n) => n);

canvas.pause(1);
mergeSort(unsorted);
